### PR TITLE
feat: create rule for $ref validation

### PR DESCRIPTION
**Pull request by IgorKarpiuk**
## What/Why/How?

Add new rule to restrict the use of $ref

## Reference

Closes: https://github.com/Redocly/redocly-cli/issues/682

## Screenshots (optional)
Знімок екрана 2023-05-31 о 13 02 12

## Check yourself

- [x] Code is linted
- [ ] Tested with redoc/reference-docs/workflows
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: packages/core/src/rules/oas3/spec-ref-validation.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+export const SpecRefValidation: Oas3Rule = () => {
+  function validateNoRef(node: unknown, { report, rawLocation }: UserContext) {
+    if (isRef(node)) {
+      report({
+        message: 'Invalid usage of $ref',
+        location: rawLocation.child('$ref').key(),
+      });
+    }
+  }
```
Comment (tatomyr) [inline packages/core/src/rules/oas3/spec-ref-validation.ts:13]: Do we need a separate function? Why don't put it simply in place [here](https://github.com/Redocly/redocly-cli/pull/1112/files?diff=unified&w=1#diff-7a126114b978592806f48c748aecefa06259fc940e41d8f3dcdc410db5380b44R44)?

File: packages/core/src/rules/oas3/spec-ref-validation.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { Oas3Rule, VisitFunction } from '../../visitors';
+import { isRef } from '../../ref-utils';
+import { UserContext } from '../../walk';
+
+export const SpecRefValidation: Oas3Rule = () => {
+  function validateNoRef(node: unknown, { report, rawLocation }: UserContext) {
+    if (isRef(node)) {
+      report({
+        message: 'Invalid usage of $ref',
```
Comment (tatomyr) [inline packages/core/src/rules/oas3/spec-ref-validation.ts:9]: What about changing the message to `Field $ref is not expected here.`?

**Review by tatomyr — COMMENTED** [review]
Is this rule only applicable to OAS3? Can we have a similar one for OAS2?

Also, you are only assessing nodes that are not allowed to contain refs. What if a $ref is somewhere inside these nodes? Wouldn't it be better to go through all nodes and report a problem when the place is not explicitly allowed to contain refs?

**Comment by IgorKarpiuk** [comment]
@tatomyr I believe we can use it for both

File: packages/core/src/rules/common/spec-ref-validation.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    'Header',
+    'SecurityScheme',
+    'Link',
+    'Callback',
+    'PathItem',
+  ];
+
+  return {
+    any(_node, { report, rawNode, rawLocation, type }) {
+      if (!nodesToValidate.includes(type.name) && isRef(rawNode)) {
```
Comment (tatomyr) [inline packages/core/src/rules/common/spec-ref-validation.ts:20]: It's hard to read :) Could you avoid negation in the statement?

Comment (IgorKarpiuk) [reply packages/core/src/rules/common/spec-ref-validation.ts:20]: I think it's better to have one condition here

File: packages/core/src/rules/common/spec-ref-validation.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    'Header',
+    'SecurityScheme',
+    'Link',
+    'Callback',
+    'PathItem',
+  ];
+
+  return {
+    any(_node, { report, rawNode, rawLocation, type }) {
+      const isShouldCheck = !nodesToSkip.includes(type.name);
```
Comment (tatomyr) [inline packages/core/src/rules/common/spec-ref-validation.ts:20]:
```suggestion
      const shouldCheck = !nodesToSkip.includes(type.name);
```

File: packages/core/src/rules/common/spec-ref-validation.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+      if (isShouldCheck && isRef(rawNode)) {
```
Comment (tatomyr) [inline packages/core/src/rules/common/spec-ref-validation.ts:22]:
```suggestion
      if (shouldCheck && isRef(rawNode)) {
```

**Review by tatomyr — APPROVED** [review]
Suggested one small improvement.
Overall looks good!

File: packages/core/src/rules/common/spec-ref-validation.ts
Comment (RomanHotsiy) [inline packages/core/src/rules/common/spec-ref-validation.ts:1]: can we still rename this rule? I would prefer `spec-strict-refs` or something esle.

`spec-ref-validation` is not consistent with our other rule names. Also, `validation` is redundant in lint rule context.

cc @IgorKarpiuk @tatomyr @adamaltman

Comment (IgorKarpiuk) [reply packages/core/src/rules/common/spec-ref-validation.ts:1]: Okay, I will rename

Comment (tatomyr) [reply packages/core/src/rules/common/spec-ref-validation.ts:1]: I like `spec-strict-refs`.

Comment (lornajane) [reply packages/core/src/rules/common/spec-ref-validation.ts:1]: I do too! Good catch @RomanHotsiy

**Pull request merged**